### DISCUSSION TITLE
Confusion around cwd

### DISCUSSION THREAD

**Discussion by fredrikaverpil** [mfussenegger/nvim-dap#1200] [Q&A] [answered]
Hi @mfussenegger 👋  and thanks for this amazing project! 

I'm somewhat confused over this line of code:

https://github.com/mfussenegger/nvim-dap/blob/6ae8a14828b0f3bff1721a35a1dfd604b6a933bb/lua/dap.lua#L373

The `run_adapter` function takes three arguments; `adapter`, `configuration` and `opts`. When using the [nvim-dap-go](https://github.com/leoluz/nvim-dap-go) adapter here, it seems like `cwd` is meant to exist directly on `adapter.cwd`, and not on `adapter.options.cwd`.

### What is the problem?

Technically, no problem that I am hitting, but I found this discrepancy while investigating the possibility to implement `cwd` for nvim-dap-go: https://github.com/leoluz/nvim-dap-go/issues/80

When I pass in `cwd` for nvim-dap-go, I have to specify it directly on the `adapter` table, not in the `adapter.options` table, or the debugger (dlv) will fail to launch in the correct `cwd`.

This is poor man's debug prints of the arguments given to `lua/dap.lua::run_adapter` when successfully launching the debugger in a nested Go project (specifying the `cwd`):

```lua
-- print(vim.inspect(adapter))
-- This comes from leoluz/nvim-dap-go.
{
  executable = {
    args = { "dap", "-l", "127.0.0.1:${port}" },
    command = "dlv",
    cwd = "/Users/fredrik/code/[REDACTED]/projectx",  -- NOTE: I'm looking to implement this, it does not exist on nvim-dap-go yet
    detached = true
  },
  options = {
    initialize_timeout_sec = 20
  },
  port = "${port}",
  type = "server"
}

-- print(vim.inspect(configuration))
{
  args = { "-test.run", "^TestSomething$" },
  cwd = "/Users/fredrik/code/[REDACTED]/projectx",
  mode = "test",
  name = "Neotest-golang Debugger",
  program = "/Users/fredrik/code/[REDACTED]/projectx/internal/core",
  request = "launch",
  type = "go"
}

-- print(vim.inspect(opts))
{
  after = <function 1>,
  before = <function 2>,
  filetype = "go"
}
```

So, it's quite unclear to me if nvim-dap maybe expects the `cwd` to be defined in both `adapter.cwd` and `adapter.options.cwd`, or whether there might actually be a bug in nvim-dap? Perhaps this is what it should say?

```diff
local function run_adapter(adapter, configuration, opts)
  local name = configuration.name or '[no name]'
  local options = adapter.options or {}
  opts = vim.tbl_extend('keep', opts, {
-    cwd = options.cwd,
+    cwd = adapter.cwd,
    env = options.env
  })
```

To top it all off, all of this was invoked by `require("neotest").run.run({ suite = false, strategy = "dap" })`, so there's also a `cwd` on the `configuration.cwd`. Even if this is likely not applicable for the _adapter_ it's somewhat confusing that there are three places from which nvim-dap could _potentially_ grab the `cwd` from.

Anyway, I'm not facing a problem here. I just figured I'd raise a concern around this being a bit confusing.

**Comment by mfussenegger** [comment, marked as answer]
I think you might be mixing up adapters of type `executable` with adapters of type `server`.
See `:help dap-adapter`.

For `executable`, there is a `options` table with `cwd`.
For `server`, there is a optional `executable` table with `cwd`. The `options` refer to the connect option. Strictly speaking for symmetry there would need to be `server.executable.options.cwd`, but I ommitted that extra level.

The code to look at is here: https://github.com/mfussenegger/nvim-dap/blob/6ae8a14828b0f3bff1721a35a1dfd604b6a933bb/lua/dap/session.lua#L1179

**Reply by fredrikaverpil** [reply]
Ah, I see. Thank you for the clarification! 😊👍
